src/bdap/linkmanager.cpp

```diff
@@ -0,0 +1,977 @@
+{
+    if (!pwalletMain)
+        return false;
+
+    std::vector<std::vector<unsigned char>> vvchMyDHTPubKeys;
+    if (!pwalletMain->GetDHTPubKeys(vvchMyDHTPubKeys))
+        return false;
+
+    if (vvchMyDHTPubKeys.size() == 0)
+        return false;
+
+    for (const std::vector<unsigned char>& vchMyDHTPubKey : vvchMyDHTPubKeys) {
+        CKeyID keyID(Hash160(vchMyDHTPubKey.begin(), vchMyDHTPubKey.end()));
+        CKeyEd25519 dhtKey;
+        if (pwalletMain->GetDHTKey(keyID, dhtKey)) {
+            std::vector<unsigned char> vchGetSharedPubKey = GetLinkSharedPubKey(dhtKey, vchLinkPubKey);
+            if (vchGetSharedPubKey == vchSharedPubKey)
+                return true;
+        }
+    }
+
+    return false;
+}
+
+bool CLinkManager::GetLinkPrivateKey(const std::vector<unsigned char>& vchSenderPubKey, const std::vector<unsigned char>& vchSharedPubKey, std::array<char, 32>& sharedSeed, std::string& strErrorMessage)
+{
+    if (!pwalletMain)
+        return false;
+
+    std::vector<std::vector<unsigned char>> vvchDHTPubKeys;
+    if (!pwalletMain->GetDHTPubKeys(vvchDHTPubKeys)) {
+        strErrorMessage = "Failed to get DHT key vector.";
+        return false;
+    }
+    // loop through each account key to check if it matches the shared key
+    for (const std::vector<unsigned char>& vchPubKey : vvchDHTPubKeys) {
+        CDomainEntry entry;
+        if (pDomainEntryDB->ReadDomainEntryPubKey(vchPubKey, entry)) {
+            CKeyEd25519 dhtKey;
+            CKeyID keyID(Hash160(vchPubKey.begin(), vchPubKey.end()));
+            if (pwalletMain->GetDHTKey(keyID, dhtKey)) {
+                if (vchSharedPubKey == GetLinkSharedPubKey(dhtKey, vchSenderPubKey)) {
+                    sharedSeed = GetLinkSharedPrivateKey(dhtKey, vchSenderPubKey);
+                    return true;
+                }
+            }
+            else {
+                strErrorMessage = strErrorMessage + "Error getting DHT private key.\n";
+            }
+        }
+    }
+    return false;
+}
+
+bool CLinkManager::FindLink(const uint256& id, CLink& link)
+{
+    if (m_Links.count(id) > 0) {
+        link = m_Links.at(id);
+        return true;
+    }
+    return false;
+}
+
+bool CLinkManager::FindLinkBySubjectID(const uint256& subjectID, CLink& getLink)
+{
+    for (const std::pair<uint256, CLink>& link : m_Links)
+    {
+        if (link.second.SubjectID == subjectID) // pending request
+        {
+            getLink = link.second;
+            return true;
+        }
+    }
+    return false;
+}
+
+void CLinkManager::ProcessQueue()
+{
+    if (!pwalletMain)
+        return;
+
+    if (pwalletMain->IsLocked())
+        return;
+
+    // make sure we are not stuck in an infinite loop
+    size_t size = QueueSize();
+    size_t counter = 0;
+    LogPrintf("CLinkManager::%s -- Start links in queue = %d\n", __func__, size);
+    while (!linkQueue.empty() && size > counter)
+    {
+        // TODO (BDAP): Do we need to lock the queue while processing?
```

Comment:
I looked into it and this code should be okay since it uses a standard queue (std::queue) instead of a standard map or multi-map.  All new items should be inserted to the back of the queue and shouldn't be touched while executing this loop.  I will remove the comment in a cleanup commit after merging this PR.